src/bootstrap/builder.rs

```diff
@@ -535,6 +526,31 @@ impl<'a> Builder<'a> {
         Some(help)
     }
 
+    fn get_builder(build: &Build, kind: Kind, paths: Vec<PathBuf>) -> Builder<'_> {
+        let top_stage = if let Some(explicit_stage) = build.config.stage {
+            explicit_stage
+        } else {
+            // See https://github.com/rust-lang/compiler-team/issues/326
+            match kind {
+                Kind::Doc => 0,
+                Kind::Build | Kind::Test => 1,
+                Kind::Bench | Kind::Dist | Kind::Install => 2,
+                // These are all bootstrap tools, which don't depend on the compiler.
+                // The stage we pass shouldn't matter, but use 0 just in case.
+                Kind::Check | Kind::Clippy | Kind::Fix | Kind::Run | Kind::Format => 0,
+            }
+        };
+        Builder {
```

Comment:
I would expect the assert to go here I think -- does that work? Something like `if on_ci { assert_eq!(top_stage, 2); }`. Obviously we might have fallout and need to add some more "oh `x.py check` is okay" or so, I'm fine with that.

Reply:
Oh that's a much better idea 😆 I implemented that.